File: FastAPI-Server/auth_service.py
import bcrypt
import jwt
from datetime import datetime, timedelta
from typing import Tuple
from uuid import uuid4
from sqlalchemy.orm import Session
from models import User, RefreshToken
from config import settings
import re

def parse_expiration_time(expires_in: str) -> timedelta:
    """Parse expiration time string like '15m', '7d' to timedelta"""
    match = re.match(r'(\d+)([mhd])', expires_in.lower())
    if not match:
        raise ValueError(f"Invalid expiration format: {expires_in}")
    
    amount, unit = int(match.group(1)), match.group(2)
    if unit == 'm':
        return timedelta(minutes=amount)
    elif unit == 'h':
        return timedelta(hours=amount)
    elif unit == 'd':
        return timedelta(days=amount)
    else:
        raise ValueError(f"Unknown time unit: {unit}")

def hash_password(password: str) -> str:
    """Hash password using bcrypt"""
    salt = bcrypt.gensalt(rounds=12)
    return bcrypt.hashpw(password.encode('utf-8'), salt).decode('utf-8')

def verify_password(password: str, password_hash: str) -> bool:
    """Verify password against hash"""
    return bcrypt.checkpw(password.encode('utf-8'), password_hash.encode('utf-8'))

def hash_token(token: str) -> str:
    """Hash token using bcrypt"""
    salt = bcrypt.gensalt(rounds=10)
    return bcrypt.hashpw(token.encode('utf-8'), salt).decode('utf-8')

def generate_access_token(user_id: int, email: str) -> str:
    """Generate access JWT token"""
    payload = {
        'sub': str(user_id),
        'email': email,
        'exp': datetime.utcnow() + parse_expiration_time(settings.JWT_ACCESS_EXPIRES_IN)
    }
    return jwt.encode(payload, settings.JWT_ACCESS_SECRET, algorithm='HS256')

def generate_refresh_token(user_id: int) -> Tuple[str, str, datetime]:
    """Generate refresh JWT token with JTI"""
    jti = str(uuid4())
    expires_in = parse_expiration_time(settings.JWT_REFRESH_EXPIRES_IN)
    expires_at = datetime.utcnow() + expires_in
    
    payload = {
        'sub': str(user_id),
        'jti': jti,
        'exp': expires_at
    }
    token = jwt.encode(payload, settings.JWT_REFRESH_SECRET, algorithm='HS256')
    return token, jti, expires_at

def persist_refresh_token(db: Session, user_id: int, token: str, jti: str, expires_at: datetime):
    """Store refresh token hash in database"""
    token_hash = hash_token(token)
    refresh_token = RefreshToken(
        jti=jti,
        token_hash=token_hash,
        user_id=user_id,
        expires_at=expires_at
    )
    db.add(refresh_token)
    db.commit()

def register_user(db: Session, email: str, password: str):
    """Register new user"""
    # Check if user already exists
    existing_user = db.query(User).filter(User.email == email).first()
    if existing_user:
        raise ValueError("Email already in use")
    
    # Create new user
    password_hash = hash_password(password)
    user = User(email=email, password_hash=password_hash)
    db.add(user)
    db.commit()
    db.refresh(user)
    
    # Generate tokens
    access_token = generate_access_token(user.id, user.email)
    refresh_token, jti, expires_at = generate_refresh_token(user.id)
    persist_refresh_token(db, user.id, refresh_token, jti, expires_at)
    
    return {
        'user': {'id': user.id, 'email': user.email},
        'tokens': {'accessToken': access_token, 'refreshToken': refresh_token}
    }

def login_user(db: Session, email: str, password: str):
    """Login user with email and password"""
    user = db.query(User).filter(User.email == email).first()
    if not user or not verify_password(password, user.password_hash):
        raise ValueError("Invalid credentials")
    
    # Generate tokens
    access_token = generate_access_token(user.id, user.email)
    refresh_token, jti, expires_at = generate_refresh_token(user.id)
    persist_refresh_token(db, user.id, refresh_token, jti, expires_at)
    
    return {
        'user': {'id': user.id, 'email': user.email},
        'tokens': {'accessToken': access_token, 'refreshToken': refresh_token}
    }

def refresh_session(db: Session, refresh_token: str):
    """Refresh user session with refresh token"""
    try:
        payload = jwt.decode(refresh_token, settings.JWT_REFRESH_SECRET, algorithms=['HS256'])
    except jwt.InvalidTokenError:
        raise ValueError("Invalid refresh token")
    
    user_id = payload.get('sub')
    jti = payload.get('jti')
    
    if not user_id or not jti:
        raise ValueError("Invalid refresh token payload")
    
    # Check if token exists and is not revoked
    token_record = db.query(RefreshToken).filter(RefreshToken.jti == jti).first()
    if not token_record or token_record.revoked:
        raise ValueError("Refresh token revoked or missing")
    
    user = db.query(User).filter(User.id == int(user_id)).first()
    if not user:
        raise ValueError("User not found")
    
    # Generate new access token
    access_token = generate_access_token(user.id, user.email)
    
    return {
        'user': {'id': user.id, 'email': user.email},
        'tokens': {'accessToken': access_token, 'refreshToken': refresh_token}
    }

def revoke_tokens_for_user(db: Session, user_id: int):
    """Revoke all refresh tokens for a user"""
    db.query(RefreshToken).filter(RefreshToken.user_id == user_id).update({'revoked': True})
    db.commit()

File: src/modules/auth/auth.service.ts
import bcrypt from "bcryptjs";
import { decode, sign, verify, Secret, SignOptions, JwtPayload as BaseJwtPayload } from "jsonwebtoken";
import { prisma } from "../../config/prisma";
import { env } from "../../config/env";
import { HttpError } from "../../middleware/errorHandler";
import { randomUUID } from "crypto";

interface JwtPayload extends BaseJwtPayload {
  sub: string;
  jti: string;
}

const accessSecret: Secret = env.jwtAccessSecret;
const refreshSecret: Secret = env.jwtRefreshSecret;
const accessExpiresIn: SignOptions["expiresIn"] = env.jwtAccessExpiresIn as SignOptions["expiresIn"];
const refreshExpiresIn: SignOptions["expiresIn"] = env.jwtRefreshExpiresIn as SignOptions["expiresIn"];

const hashToken = async (token: string): Promise<string> => bcrypt.hash(token, 10);

const generateAccessToken = (userId: number, email: string): string => {
  return sign({ sub: String(userId), email }, accessSecret, {
    expiresIn: accessExpiresIn,
  });
};

const generateRefreshToken = async (userId: number): Promise<{ token: string; jti: string; expiresAt: Date }> => {
  const jti = randomUUID();
  const token = sign({ sub: String(userId), jti }, refreshSecret, {
    expiresIn: refreshExpiresIn,
  });

  // Decode expiresAt from token
  const decoded = decode(token) as { exp?: number } | null;
  const expiresAt = decoded?.exp ? new Date(decoded.exp * 1000) : new Date(Date.now() + 7 * 24 * 60 * 60 * 1000);

  return { token, jti, expiresAt };
};

const persistRefreshToken = async (userId: number, token: string, jti: string, expiresAt: Date) => {
  const tokenHash = await hashToken(token);
  await prisma.refreshToken.create({
    data: { userId, jti, tokenHash, expiresAt },
  });
};

export const registerUser = async (email: string, password: string) => {
  const existing = await prisma.user.findUnique({ where: { email } });
  if (existing) {
    throw new HttpError(409, "Email already in use");
  }

  const passwordHash = await bcrypt.hash(password, 12);
  const user = await prisma.user.create({ data: { email, passwordHash } });

  const accessToken = generateAccessToken(user.id, user.email);
  const { token: refreshToken, jti, expiresAt } = await generateRefreshToken(user.id);
  await persistRefreshToken(user.id, refreshToken, jti, expiresAt);

  return {
    user: { id: user.id, email: user.email },
    tokens: { accessToken, refreshToken },
  };
};

export const loginUser = async (email: string, password: string) => {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) {
    throw new HttpError(401, "Invalid credentials");
  }

  const valid = await bcrypt.compare(password, user.passwordHash);
  if (!valid) {
    throw new HttpError(401, "Invalid credentials");
  }

  const accessToken = generateAccessToken(user.id, user.email);
  const { token: refreshToken, jti, expiresAt } = await generateRefreshToken(user.id);
  await persistRefreshToken(user.id, refreshToken, jti, expiresAt);

  return {
    user: { id: user.id, email: user.email },
    tokens: { accessToken, refreshToken },
  };
};

export const refreshSession = async (refreshToken: string) => {
  let payload: JwtPayload;
  try {
    payload = verify(refreshToken, refreshSecret) as JwtPayload;
  } catch {
    throw new HttpError(401, "Invalid refresh token");
  }

  if (!payload?.sub || !payload?.jti) {
    throw new HttpError(401, "Invalid refresh token payload");
  }

  const tokenRecord = await prisma.refreshToken.findUnique({ where: { jti: payload.jti } });
  if (!tokenRecord || tokenRecord.revoked) {
    throw new HttpError(401, "Refresh token revoked or missing");
  }

  const stillValid = await bcrypt.compare(refreshToken, tokenRecord.tokenHash);
  if (!stillValid) {
    throw new HttpError(401, "Invalid refresh token");
  }

  if (tokenRecord.expiresAt < new Date()) {
    throw new HttpError(401, "Refresh token expired");
  }

  // rotate token
  await prisma.refreshToken.update({ where: { jti: payload.jti }, data: { revoked: true } });

  const userId = Number(payload.sub);
  const user = await prisma.user.findUnique({ where: { id: userId } });
  if (!user) {
    throw new HttpError(404, "User not found");
  }

  const accessToken = generateAccessToken(user.id, user.email);
  const { token: newRefreshToken, jti, expiresAt } = await generateRefreshToken(user.id);
  await persistRefreshToken(user.id, newRefreshToken, jti, expiresAt);

  return {
    user: { id: user.id, email: user.email },
    tokens: { accessToken, refreshToken: newRefreshToken },
  };
};

export const revokeTokensForUser = async (userId: number) => {
  await prisma.refreshToken.updateMany({ where: { userId, revoked: false }, data: { revoked: true } });
};
